src/Aspire.Hosting.Analyzers/AppHostAnalyzer.cs

```diff
@@ -0,0 +1,147 @@
+        ModelType modelType = default;
+        var candidateParameter = targetMethod.Parameters.FirstOrDefault(ps =>
+            SymbolEqualityComparer.Default.Equals(ps.Type, wellKnownTypes.Get(SpecialType.System_String))
+            && HasModelNameAttribute(ps, out modelType));
+
+        if (candidateParameter is not null)
+        {
+            parameterData = (candidateParameter, modelType);
+            return true;
+        }
+
+        parameterData = null;
+        return false;
+
+        bool HasModelNameAttribute(IParameterSymbol parameter, out ModelType modelType)
+        {
+            var modelNameParameter = wellKnownTypes.Get(WellKnownTypeData.WellKnownType.Aspire_Hosting_ApplicationModel_IModelNameParameter);
+            var resourceNameAttribute = wellKnownTypes.Get(WellKnownTypeData.WellKnownType.Aspire_Hosting_ApplicationModel_ResourceNameAttribute);
+            var endpointNameAttribute = wellKnownTypes.Get(WellKnownTypeData.WellKnownType.Aspire_Hosting_ApplicationModel_EndpointNameAttribute);
+
+            var attrData = parameter.GetAttributes().SingleOrDefault(a => WellKnownTypes.Implements(a.AttributeClass, modelNameParameter));
+
+            // Model type (e.g. Resource, Endpoint) is based on the concrete attribute type
+            modelType = SymbolEqualityComparer.Default.Equals(attrData?.AttributeClass, resourceNameAttribute)
+                ? ModelType.Resource
+                : SymbolEqualityComparer.Default.Equals(attrData?.AttributeClass, endpointNameAttribute)
+                  ? ModelType.Endpoint
+                  : ModelType.Unknown;
+
+            return attrData is not null;
+        }
+    }
+
+    private static bool TryGetStringToken(IInvocationOperation invocation, IParameterSymbol modelNameParameter, out SyntaxToken token)
+    {
+        IArgumentOperation? argumentOperation = null;
+        foreach (var argument in invocation.Arguments)
+        {
+            if (SymbolEqualityComparer.Default.Equals(modelNameParameter, argument.Parameter))
+            {
+                argumentOperation = argument;
+                break;
+            }
+        }
+
+        if (argumentOperation?.Syntax is not ArgumentSyntax routePatternArgumentSyntax ||
+            routePatternArgumentSyntax.Expression is not LiteralExpressionSyntax routePatternArgumentLiteralSyntax)
+        {
+            token = default;
+            return false;
+        }
+
+        token = routePatternArgumentLiteralSyntax.Token;
+        return true;
+    }
+
+    private record struct ModelNameOperation(IInvocationOperation Operation, ModelType ModelType, SyntaxToken ModelNameToken)
+    {
+        public static ModelNameOperation Create(IInvocationOperation operation, ModelType modelType, SyntaxToken modelNameToken)
+        {
+            return new ModelNameOperation(operation, modelType, modelNameToken);
+        }
+    }
+
+    private enum ModelType
+    {
+        Unknown,
+        Resource,
+        Endpoint
```

Comment:
Where is `ModelType.Endpoint` used? Can we remove this and the attribute until it is used? We can easily add it back when we need it.

Reply:
Yeah it's not actually used currently, I can remove.

Reply:
OK it *is* used it turns out, in that the model type is passed as string to `ModelName.TryValidateName`. That's why this is there.